lua/neogit/lib/buffer.lua

```diff
@@ -139,8 +139,38 @@ function Buffer:set_extmarks(extmarks)
 end
 
 function Buffer:set_line_highlights(highlights)
+  local line_ansi_colorized_map = {}
+
   for _, hl in ipairs(highlights) do
-    self:add_line_highlight(unpack(hl))
+    local line_nr, hl_group = unpack(hl)
+    if hl_group == "NeogitDiffContext" then
+      line_ansi_colorized_map[line_nr] = true
+    else
+      self:add_line_highlight(unpack(hl))
+    end
+  end
+
+  local line_ansi_colorized = {}
+  for k in pairs(line_ansi_colorized_map) do
+    table.insert(line_ansi_colorized, k)
+  end
+  table.sort(line_ansi_colorized)
+
+  local start_line_nr, prev_line_nr
+  for _, line_nr in ipairs(line_ansi_colorized) do
+    if start_line_nr == nil then
+      start_line_nr = line_nr
+    end
+    if prev_line_nr ~= nil and line_nr ~= prev_line_nr + 1 then
+      local text = self:get_lines(start_line_nr, prev_line_nr + 1, false)
+      vim.g.baleia.buf_set_lines(self.handle, start_line_nr, prev_line_nr + 1, false, text)
+      start_line_nr = line_nr
+    end
+    prev_line_nr = line_nr
+  end
+  if start_line_nr ~= nil then
+    local text = self:get_lines(start_line_nr, prev_line_nr + 1, false)
+    vim.g.baleia.buf_set_lines(self.handle, start_line_nr, prev_line_nr + 1, false, text)
```

Comment:
My feeling is that this is leaking too much higher-level knowledge of the content passed into a simple setter method, plus the hardcoded "NeogitDiffContext" highlight group seems like something easy to break on accident.

Couple ideas:
- Would it be possible to use the ansi parser I wrote for the git graphs for this task instead of baleia? (https://github.com/NeogitOrg/neogit/blob/master/lua/neogit/lib/ansi.lua). It would need to be extended to handle more codes, but that's doable. 
- If that's possible, we could push up the parsing to the view layer, leaving this method alone.

Reply:
I like this, but it seems like a lot of work without some library.

I had a look at `neogit/lua/neogit/lib/ansi.lua`. This matches some ANSI SGR code, then some ANSI attribute reset. ANSI codes don't have to obey this pattern though. An example output from *Delta*:

    "\27[34m\27[38;2;119;119;119m 513\27[34m┊\27[0m                                                     \27[34m\27[38;2;119;119;119m 513\27[34m│\27[0m"

The ANSI codes here include:

- `\27[34m`: Set foreground to blue.
- `\27[38;2;119;119;119m`: Set foreground to `rgb(119, 119, 119)`
- `\27[0m`: Reset.

Coupled with styles like bold, faint, italic, underline, ..., I feel like I'd have to reimplement most of *baleia*.

Reply:
I removed the hardcoded `NeogitDiffContext` now in the commit <https://github.com/NeogitOrg/neogit/pull/1593/commits/5bfc499b9c51b01663aefd49cc05e20cf0d26b14>.

Reply:
Yea, totally fair. It's very custom made for parsing git-log graph output, so, not really surprising that it's not sufficient. TBH I kinda wish I had found baleia when I made that, it was a real pain getting that pattern right.

Anyways, iirc the terminal buffer has some other semantics and might be tougher to integrate with the component rendering. 

Looking at where it's at now, it feels very nicely integrated with the existing system, so I don't have any more notes. Nice work :)